src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,950 @@
+    /// ```
+    ///
+    /// Then `wake_up_delay_duration` is elapsed since the previous waking up.
+    /// Here, we expect that the lock wait entries 30 and 40 can be woken
+    /// up, since they exists when the previous waking up occurs. But we
+    /// don't want to wake up later-arrived entries (50 and 60) since it
+    /// introduces useless pessimistic retries to transaction 50 and 60 when
+    /// they don't need to. The solution is, only wake up the entries that
+    /// has `entry.current_legacy_wake_up_cnt <
+    /// key_lock_wait_state.legacy_wake_up_cnt`. Therefore, we only wakes up
+    /// entries 30 and 40 who has `current_legacy_wake_up_cnt < 1`, while 50 and
+    /// 60 will be left untouched.
+    ///
+    /// When waking up resumable requests, the mechanism above won't take
+    /// effect. If a legacy request is woken up and triggered the mechanism,
+    /// and there is a resumable request in the queue, `delayed_notify_all`
+    /// will stop at the first resumable request it meets, pop it out, and
+    /// return it from a [`DelayedNotifyAllFuture`]. See
+    /// [`LockWaitQueues::pop_for_waking_up`].
+    legacy_wake_up_cnt: usize,
+    queue: BinaryHeap<Box<LockWaitEntry>>,
+
+    /// The start_ts of the most recent waking up event.
+    last_conflict_start_ts: TimeStamp,
+    /// The commit_ts of the most recent waking up event.
+    last_conflict_commit_ts: TimeStamp,
+
+    /// `(id, start_time, delay_duration)`
+    delayed_notify_all_state: Option<(u64, Instant, Arc<AtomicU64>)>,
+}
+
+impl KeyLockWaitState {
+    fn new() -> Self {
+        Self {
+            current_lock: kvrpcpb::LockInfo::default(),
+            legacy_wake_up_cnt: 0,
+            queue: BinaryHeap::new(),
+            last_conflict_start_ts: TimeStamp::zero(),
+            last_conflict_commit_ts: TimeStamp::zero(),
+            delayed_notify_all_state: None,
+        }
+    }
+}
+
+pub type DelayedNotifyAllFuture = Pin<Box<dyn Future<Output = Option<Box<LockWaitEntry>>> + Send>>;
+
+pub struct LockWaitQueueInner {
+    queue_map: dashmap::DashMap<Key, KeyLockWaitState>,
+    id_allocated: AtomicU64,
+}
+
+#[derive(Clone)]
+pub struct LockWaitQueues<L: LockManager> {
+    inner: Arc<LockWaitQueueInner>,
+    #[allow(dead_code)]
+    lock_mgr: L,
+}
+
+impl<L: LockManager> LockWaitQueues<L> {
+    pub fn new(lock_mgr: L) -> Self {
+        Self {
+            inner: Arc::new(LockWaitQueueInner {
+                queue_map: dashmap::DashMap::new(),
+                id_allocated: AtomicU64::new(1),
+            }),
+            lock_mgr,
+        }
+    }
+
+    /// Enqueues a lock wait entry. The key is indicated by the `key` field of
+    /// the `lock_wait_entry`. The caller also need to provide the
+    /// information of the current-holding lock.
+    pub fn push_lock_wait(
+        &self,
+        mut lock_wait_entry: Box<LockWaitEntry>,
+        current_lock: kvrpcpb::LockInfo,
+    ) {
+        let mut entry = self
```

Comment:
`entry` is a common name of the elements in a Map, and it's easy to confuse with `LockWaitEntry`. 
um...,  it maybe ok ?

Reply:
Maybe there can be some better name for  `LockWaitEntry?
cc @sticnarf , Is there any suggested name?

Reply:
My intention is to rename the temporary variable `entry` to a name relative to its meanings, like `lock_wait_state` or else.

Reply:
Yep, I also agree the local variable `entry` can be changed to `state because `dashmap::Entry` can be deref as `KeyLockWaitState`. I don't have any good name suggestion for `LockWaitEntry`...